design/ceph/nvmeof-recoverer.md

```diff
@@ -0,0 +1,184 @@
+
+- **Load balance**: As storage needs grow, Recoverer enables easy expansion of storage capacity. It introduces a scheduler that automatically connects fabric devices to suitable nodes using load-balancing and deploys OSDs. By utilizing Kubernetes Jobs, the scheduler dynamically connects based on the workload ensuring optimal performance
+
+### Non-Goals
+
+- **Support for Non-NVMe-oF Environments**: This design focuses solely on NVMe-oF environments.
+- **Modifications to the Core Ceph Codebase**: The proposal does not involve changing Ceph's core functionalities.
+
+## Proposal Details
+
+### Overview
+
+We propose a design, called NVMe-oF recoverer. NVMe-oF is an emerging storage network where the storage devices can be attached to any storage node using RDMA, TCP or Fiber channel. In this environment, the failure caused by a server node can be recoverable because the OSD can be re-spawned to another server node with the storage device attached to the failed server node through NVMe-oF.
+
+To address these challenges, we recommend a phased implementation consisting of four key steps:
+
+1. **Implement an Explicit Command for Manual OSD Relocation**
+2. **Automate OSD Relocation Upon Failure**
+3. **Modify CRUSH Map to Minimize Data Rebalancing**
+4. **Improve OSD Scheduling Based on Node Workloads**
+
+Each step can be implemented and contributed individually, allowing for incremental progress and testing.
+
+### Step 1: Implement an Explicit Command for Manual OSD Relocation
+
+#### Description
+
+Introduce explicit commands or scripts that allow administrators to manually deploy and relocate OSD pods using NVMe-oF devices. These tools will streamline the management of NVMe-oF devices and OSD pods in standard scenarios and during failures.
+
+#### Implementation Steps
+
+- **Develop a Script or Command-line Tool**:
+  - **For Connecting NVMe-oF Devices and Deploying OSD Pods**:
+    - Retrieve NVMe-oF target information from predefined sources (e.g., configuration files, environment variables).
+      - Key parameters: IP, Port, Subsystem NQN (SubNQN), and attachable nodes.
+    - Connect devices to target nodes using `nvme-cli` or appropriate tools.
+    - Update the `CephCluster` Custom Resource Definition (CRD) with fabric device details to facilitate automated OSD deployment.
+  - **For Handling OSD Relocation in Failure Scenarios**:
+    - Remove the faulted OSD pod.
+    - Disconnect the device from the failed node and reconnect it to a healthy node.
+    - Update the `CephCluster` CRD to reflect the new device-node mapping.
```

Comment:
Is this only updating the status? Rook only updates the CR status. Only users are allowed to update the CR spec.